src/analysis/processing/qgsalgorithmrandompointsextent.cpp

```diff
@@ -0,0 +1,195 @@
+}
+
+bool QgsRandomPointsExtentAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsCoordinateReferenceSystem extent_crs = parameterAsExtentCrs( parameters, QStringLiteral( "EXTENT" ), context );
+  QgsCoordinateReferenceSystem user_crs = parameterAsCrs( parameters, QStringLiteral( "TARGET_CRS" ), context );
+
+  //prefer user crs over extent/project crs if defined
+  if ( extent_crs != user_crs )
+    mCrs = user_crs;
+  else
+    mCrs = extent_crs;
+
+  mExtent = parameterAsExtent( parameters, QStringLiteral( "EXTENT" ), context, mCrs );
+  mNumPoints = parameterAsInt( parameters, QStringLiteral( "POINTS_NUMBER" ), context );
+  mDistance = parameterAsDouble( parameters, QStringLiteral( "MIN_DISTANCE" ), context );
+  mMaxAttempts = parameterAsInt( parameters, QStringLiteral( "MAX_ATTEMPTS" ), context );
+
+  return true;
+}
+
+QVariantMap QgsRandomPointsExtentAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+
+  QgsFields fields = QgsFields();
+  fields.append( QgsField( QStringLiteral( "id" ), QVariant::LongLong ) );
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields, QgsWkbTypes::Point, mCrs ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  //initialize random engine
+  std::random_device random_device;
+  std::mt19937 mersenne_twister( random_device() );
+
+  std::uniform_real_distribution<double> x_distribution( mExtent.xMinimum(), mExtent.xMaximum() );
+  std::uniform_real_distribution<double> y_distribution( mExtent.yMinimum(), mExtent.yMaximum() );
+
+  if ( mDistance == 0 )
+  {
+    int i = 0;
+    while ( i < mNumPoints )
+    {
+      if ( feedback && feedback->isCanceled() )
+        break;
+
+      double rx = x_distribution( random_device );
+      double ry = y_distribution( random_device );
+
+      QgsFeature f = QgsFeature( i );
+
+      f.setGeometry( QgsGeometry( new QgsPoint( rx, ry ) ) );
+      f.setAttributes( QgsAttributes() << i );
+      sink->addFeature( f, QgsFeatureSink::FastInsert );
+      i++;
+      feedback->setProgress( static_cast<int>( static_cast<double>( i ) / static_cast<double>( mNumPoints ) * 100 ) );
+    }
+  }
+  else
+  {
+    QgsSpatialIndex index = QgsSpatialIndex();
+    int distCheckIterations = 0;
+
+    int i = 0;
+    while ( i < mNumPoints )
+    {
+      if ( feedback && feedback->isCanceled() )
+        break;
+
+      double rx = x_distribution( random_device );
+      double ry = y_distribution( random_device );
+
+      QgsGeometry randomPointGeom = QgsGeometry( new QgsPoint( rx, ry ) );
+
+      //check if new random point is inside searching distance to existing points
+      QList<QgsFeatureId> neighbors = index.nearestNeighbor( randomPointGeom, 1, mDistance );
```

Comment:
Very minor optimisation, but worthwhile:

```suggestion
      QList<QgsFeatureId> neighbors = index.nearestNeighbor( QgsPointXY( rx, ry), 1, mDistance );
```

and then move `QgsGeometry randomPointGeom = QgsGeometry( new QgsPoint( rx, ry ) );`
inside the `if ( neighbors.empty() )` branch. This avoids creating the QgsPoint on the heap whenever it violates the min distance constraint.

Reply:
Thanks @nyalldawson, I just added your remarks.